server/lib/prune-newick.js

```diff
@@ -0,0 +1,163 @@
+/*
+This file will prune a newick tree to remove genes that are too dissimilar. 
+*/
+
+const hammingDistance = require('../hamdis/hamming-distance')
+const {parseFasta} = require('../formats/fasta/parse')
+
+module.exports.pruneOutliers = pruneOutliers
+
+function pruneOutliers(newick, alignedFasta){
+
+	const fastaData = parseFasta(alignedFasta)
+	// Build a dict map of species -> sequence 
+	let sequenceMap = {}
+	for(let obj of fastaData){
+		sequenceMap[obj.species] = obj.sequence
+	}
+
+	let leafNodes = []
+	function GetLeaves(node) {
+		if (node.branchset) {
+			node.branchset.forEach(GetLeaves)
+		} else {
+			leafNodes.push(node)
+		}
+	}
+	// Compute average and standard deviation of all the pairs of distances 
+	GetLeaves(newick)
+	let average = 0
+	let totalDistances = []
+	let distCache = {}
+	
+	for(let i=0;i < leafNodes.length; i++){
+		let node = leafNodes[i]
+		let species1 = node.ident ? (node.name + '__' + node.ident) : node.name
+		distCache[i] = {}
+		for(let j=0;j < leafNodes.length;j++){
+			if(i == j){
+				continue
+			}
+
+			let species2 = leafNodes[j].ident ? (leafNodes[j].name + '__' + leafNodes[j].ident) : leafNodes[j].name
+			let dist = hammingDistance(sequenceMap[species1],sequenceMap[species2])
+			average += dist 
+			totalDistances.push(dist)
+			distCache[i][j] = dist
+		}
+	}
+	average /= totalDistances.length 
+
+	let std = 0;
+	for(let dist of totalDistances) {
+		std += Math.pow((dist - average),2)
+	}
+	std /= totalDistances.length
+	std = Math.sqrt(std)
+
+	let toRemoveNames = []
+	let toRemoveNodes = []
+
+	// Remove things if a majority of their pair checks are above average  
+	for(let i=0;i < leafNodes.length; i++){
+		let node = leafNodes[i]
+		let diffCount = 0
+
+		for(let j=0;j < leafNodes.length;j++){
```

Comment:
Ditto here, `let [j, _node] of leafNodes.entries()`? (prefix `_node` because value is not used)